popup/search-results.js

```diff
@@ -0,0 +1,547 @@
+      $('#searchResultsNav-prev').setAttribute('disabled', 'disabled');
+    } else {
+      $('#searchResultsNav-prev').removeAttribute('disabled');
+    }
+    $('#searchResultsNav-currentPage').textContent = currentDisplayedPage;
+
+    let totalResultsCount = processedResults.length;
+    if (unprocessedResults.length > 0) {
+      // Add 1 page if there's results left to process.
+      totalResultsCount += DISPLAYED_RESULTS_PER_PAGE;
+    }
+    const totalPageCount = Math.ceil(Math.max(1, totalResultsCount / DISPLAYED_RESULTS_PER_PAGE));
+    if (currentDisplayedPage >= totalPageCount || loading) {
+      $('#searchResultsNav-next').setAttribute('disabled', 'disabled');
+    } else {
+      $('#searchResultsNav-next').removeAttribute('disabled');
+    }
+    $('#searchResultsNav-totalPages').textContent = totalPageCount;
+
+    const navNode = $('#searchResultsNav');
+    if (loading && !navNode.classList.contains('loading')) {
+      navNode.classList.add('loading');
+    } else {
+      navNode.classList.remove('loading');
+    }
+  }
+
+  /**
+   * @returns {Boolean} If we should process more results.
+   */
+  function shouldLoadMore() {
+    return (processedResults.length < currentDisplayedPage * DISPLAYED_RESULTS_PER_PAGE);
+  }
+
+  function loadMoreIfNeeded() {
+    if (shouldLoadMore()) {
+      loading = true;
+      render();
+      setTimeout(load, DELAY_BEFORE_SEARCHING_STYLES);
+    } else {
+      loading = false;
+      render();
+    }
+  }
+
+  /** Increments currentDisplayedPage and loads results. */
+  function next(event) {
+    if (event) {
+      event.preventDefault();
+    }
+    currentDisplayedPage += 1;
+    render();
+    loadMoreIfNeeded();
+  }
+
+  /** Decrements currentPage and loads results. */
+  function prev(event) {
+    if (event) {
+      event.preventDefault();
+    }
+    currentDisplayedPage = Math.max(1, currentDisplayedPage - 1);
+    render();
+  }
+
+  /**
+   * Display error message to user.
+   * @param {string} message  Message to display to user.
+   */
+  function error(reason) {
+    let message;
+    if (reason === 404) {
+      // TODO: i18n message
+      message = 'No results found';
+    } else {
+      message = 'Error loading search results: ' + reason;
+    }
+    $('#searchResults').classList.add('hidden');
+    $('#searchResults-error').innerHTML = message;
+    $('#searchResults-error').classList.remove('hidden');
+  }
+
+  /**
+   * Initializes search results container, starts fetching results.
+   * @param {Object} event The click event
+   */
+  function load(event) {
+    if (event) {
+      event.preventDefault();
+    }
+
+    loading = true;
+    render();
+
+    if (unprocessedResults.length > 0) {
+      processNextResult();
+      return true;
+    }
+
+    if (searchAPI.isExhausted()) {
+      loading = false;
+      render();
+      return true;
+    }
+
+    $('#find-styles').classList.add('hidden');
+    $('#open-search').classList.remove('hidden');
```

Comment:
While the steps look good, but the states should be grouped; move one of them so that `add`'s are together. Might even add an empty line between.